controller_manager:
  ros__parameters:
    update_rate: 100  # Hz

    # Standard Broadcasters
    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster

    # Required by real UR driver for Robot 1
    io_and_status_controller:
      type: ur_controllers/GPIOController
    speed_scaling_state_broadcaster:
      type: ur_controllers/SpeedScalingStateBroadcaster
    force_torque_sensor_broadcaster:
      type: force_torque_sensor_broadcaster/ForceTorqueSensorBroadcaster
    tcp_pose_broadcaster:
      type: pose_broadcaster/PoseBroadcaster
    ur_configuration_controller:
      type: ur_controllers/URConfigurationController

    # Trajectory Controllers for Robot 1
    scaled_joint_trajectory_controller:
      type: ur_controllers/ScaledJointTrajectoryController
    joint_trajectory_controller:
      type: joint_trajectory_controller/JointTrajectoryController

    # Trajectory Controller for Robot 2
    ur_2_controller:
      type: joint_trajectory_controller/JointTrajectoryController



    ur_2_gripper_controller:
          type: position_controllers/GripperActionController

    enforce_command_limits: false

# ----------------------------------------------------------------------
# ROBOT 1 SPECIFIC PARAMETERS (Real Robot)
# ----------------------------------------------------------------------
speed_scaling_state_broadcaster:
  ros__parameters:
    state_publish_rate: 100.0

force_torque_sensor_broadcaster:
  ros__parameters:
    sensor_name: tcp_fts_sensor
    state_interface_names:
      - force.x
      - force.y
      - force.z
      - torque.x
      - torque.y
      - torque.z
    frame_id: tool0_controller
    topic_name: ft_data

tcp_pose_broadcaster:
  ros__parameters:
    frame_id: base
    pose_name: tcp_pose
    tf:
      child_frame_id: tool0_controller

scaled_joint_trajectory_controller:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint
    command_interfaces:
      - position
    state_interfaces:
      - position
      - velocity
    state_publish_rate: 100.0
    action_monitor_rate: 20.0
    allow_partial_joints_goal: false
    speed_scaling_interface_name: speed_scaling/speed_scaling_factor
    constraints:
      stopped_velocity_tolerance: 0.2
      goal_time: 0.0

joint_trajectory_controller:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint
    command_interfaces:
      - position
    state_interfaces:
      - position
      - velocity
    state_publish_rate: 100.0
    action_monitor_rate: 20.0
    allow_partial_joints_goal: false
    constraints:
      stopped_velocity_tolerance: 0.2
      goal_time: 0.0

# ----------------------------------------------------------------------
# ROBOT 2 SPECIFIC PARAMETERS (Simulated / Secondary Robot)
# ----------------------------------------------------------------------
ur_2_controller:
  ros__parameters:
    joints:
      - ur_2_shoulder_pan_joint
      - ur_2_shoulder_lift_joint
      - ur_2_elbow_joint
      - ur_2_wrist_1_joint
      - ur_2_wrist_2_joint
      - ur_2_wrist_3_joint

    command_interfaces:
      - position
    state_interfaces:
      - position
      - velocity
    state_publish_rate: 100.0
    action_monitor_rate: 20.0
    allow_partial_joints_goal: false
    constraints:
      stopped_velocity_tolerance: 0.2
      goal_time: 0.0

ur_2_gripper_controller:
  ros__parameters:
    joint: ur_2_robotiq_85_left_knuckle_joint
    action_monitor_rate: 20.0
    goal_tolerance: 0.01
    max_effort: 100.0